Web process sends redundant information when establishing GPU process connection

kkinnunen-apple opened the pull request:
#### c42e0209c0ba6ca8df94f34d6258c66f779a633f
<pre>
Web process sends redundant information when establishing GPU process connection
<a href="https://bugs.webkit.org/show_bug.cgi?id=274307">https://bugs.webkit.org/show_bug.cgi?id=274307</a>
<a href="https://rdar.apple.com/128265618">rdar://128265618</a>

Reviewed by Chris Dumez.

Web process would send
 - the process identity for resource attribution
 - settings object

The settings were previously changled to be filled by the privileged
UI process. Remove this altogether, as it&apos;s insecure to let the WP
select anything here.

The process identity does not change once the WP is created. Thus
send this during WP initialization.

Moves the WebProcessProxy message sends to the WebProcessProxy.

* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::initializeIdentity):
(WebKit::WebProcessProxy::createGPUProcessConnection):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::create):
(WebKit::getGPUProcessConnectionParameters): Deleted.
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):

Canonical link: <a href="https://commits.webkit.org/279259@main">https://commits.webkit.org/279259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a98ab689b2f598ad5a5725b80f432ab38b2fc3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52770 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32107 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5254 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56048 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3493 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38861 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3215 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42835 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2250 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54868 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29871 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45548 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23947 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/52614 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26955 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1652 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48830 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3012 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57640 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27909 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3010 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50234 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29131 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45742 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49515 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11556 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30047 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28883 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
```diff
@@ -103,35 +102,15 @@
-    if (!connectionIdentifiers)
-        return nullptr;
-
-    RELEASE_ASSERT_WITH_MESSAGE(WebProcess::singleton().hasEverHadAnyWebPages(), "GPUProcess preferences come from the pages");
```
this assertion is now removed, as it was originally there only to say "we have settings from the pages"

cdumez commented on Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
```diff
@@ -103,35 +102,15 @@
 
-GPUProcessConnection::GPUProcessConnection(IPC::Connection::Identifier&& connectionIdentifier)
-    : m_connection(IPC::Connection::createServerConnection(connectionIdentifier))
+GPUProcessConnection::GPUProcessConnection(Ref<IPC::Connection> connection)
```
`Ref<IPC::Connection>&&` or `IPC::Connection&` but we shouldn't be passing smart pointers by value.

kkinnunen-apple replied:
Passing Ref<T> is more correct and optimal than passing Ref<T>&&.

More correct: 

communicates and forces ownership transfer (of the outstanding reference, in this case). The caller won't have any references to the instance during the call.

More optimal:

The instance may be created into the call frame
As opposed to:
The instance is created in the caller frame and then redundant pointer passed in the call frame.

https://stackoverflow.com/questions/71930850/why-not-to-accept-stdunique-ptr-by-rvalue-reference

cdumez replied:
This is not the pattern we use in WebKit. The reason being that this doesn't force the caller to WTFMove() and thus is likely to cause unnecessary ref-counting churn in practice.

cdumez commented on Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
```diff
@@ -103,35 +102,15 @@
-    if (parentConnection.ignoreInvalidMessageForTesting())
-        instance->connection().setIgnoreInvalidMessageForTesting();
-#endif
+    auto instance = adoptRef(*new GPUProcessConnection(WTFMove(connection)));
```
`auto` -> `Ref`

cdumez commented on Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
```diff
@@ -66,7 +66,7 @@ class RemoteVideoFrameObjectHeapProxy;
 class GPUProcessConnection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<GPUProcessConnection>, public IPC::Connection::Client {
 public:
-    static RefPtr<GPUProcessConnection> create(IPC::Connection& parentConnection);
+    static Ref<GPUProcessConnection> create(Ref<IPC::Connection>);
```
We should not pass smart pointer by value.

kkinnunen-apple replied:
The idea is to communicate the pass of the ownership of the reference, similar to when passing unique_ptr.

cdumez replied:
This is what `&&` is for. Please follow our patterns.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1364,9 +1366,17 @@ GPUProcessConnection& WebProcess::ensureGPUProcessConnection()
+        auto connectionIdentifiers = IPC::Connection::createConnectionIdentifierPair();
+        if (!connectionIdentifiers)
             CRASH();
+        RefPtr connection = protectedParentProcessConnection();
```
This doesn't need to call `protectedParentProcessConnection()` since you're assigning to a RefPtr. Should call parentProcessConnection().

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -413,6 +413,8 @@ static void scheduleLogMemoryStatistics(LogMemoryStatisticsReason reason)
 {    
     TraceScope traceScope(InitializeWebProcessStart, InitializeWebProcessEnd);
+    ProcessIdentity p { ProcessIdentity::CurrentProcess };
+    parentProcessConnection()->send(Messages::WebProcessProxy::InitializeIdentity(WTFMove(p)), 0);
```
Maybe we should use sendWithAsyncReply for initializeWebProcess() and have it return the process identity?

Looks to me that `WebProcessProxy::InitializeIdentity` is only sent once.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1067,13 +1074,18 @@ void WebProcessProxy::getNetworkProcessConnection(CompletionHandler<void(Network
+void WebProcessProxy::createGPUProcessConnection(IPC::Connection::Handle&& connectionIdentifier)
 {
+    WebKit::GPUProcessConnectionParameters parameters;
+    parameters.webProcessIdentity = m_processIdentity;
```
Shouldn't we `ASSERT(m_processIdentity);` before this?

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -347,6 +348,7 @@ void WebProcessProxy::platformInitialize()
 }
 #endif
 
+
```
nit: I don't think we need this newline?

cdumez reviewed (approved):
LGTM

kkinnunen-apple committed:
Web process sends redundant information when establishing GPU process connection
https://bugs.webkit.org/show_bug.cgi?id=274307
rdar://128265618

Reviewed by Chris Dumez.

Web process would send
 - the process identity for resource attribution
 - settings object

The settings were previously changled to be filled by the privileged
UI process. Remove this altogether, as it's insecure to let the WP
select anything here.

The process identity does not change once the WP is created. Thus
send this during WP initialization.

Moves the WebProcessProxy message sends to the WebProcessProxy.

* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::initializeIdentity):
(WebKit::WebProcessProxy::createGPUProcessConnection):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::create):
(WebKit::getGPUProcessConnectionParameters): Deleted.
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):

Canonical link: https://commits.webkit.org/279259@main
